### PR TITLE
Streamline honoring of “singleThreaded” attribute

### LINKED ISSUE #144: @Test(singleThreaded = true) from (abstract) parent class should apply to all childs
I've got `@Test(singleThreaded = true)` on abstract class and i expect it to be applied to the child class(es) so that each child class can be run in parallel but only with one thread each. 

Also when i add `@Test(singleThreaded = true)` to the child class it don't work either. But when i change testng.xml from

``` xml
<suite name="Suite" preserve-order="false" parallel="tests">
    <test name="Test" preserve-order="false" parallel="classes">
        <packages>
            <package name="my.package" />
        </packages>
    </test>
</suite>
```

to

``` xml
<suite name="Suite" preserve-order="false" parallel="tests">
    <test name="Test" preserve-order="false" parallel="false">
        <packages>
            <package name="my.package" />
        </packages>
    </test>
</suite>
```

tests pass! Also it's clear from logs (with thread name logging enabled) that test methods from one class are actually run in parallel.

### LINKED ISSUE #2361: No way to enforce @Test(singleThreaded = true) when test defined in base class
### TestNG Version

6.10, 7.3.0

### Expected behavior

Methods executed serially

### Actual behavior

Tests executed parallel

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### Test case sample

```java
@Test(singleThreaded = true)
public abstract class BaseTestClass
{
    private final AtomicInteger currentTests = new AtomicInteger();

    protected void test()
    {
        int currentTests = this.currentTests.incrementAndGet();
        try {
            assertEquals(currentTests, 1);
            MILLISECONDS.sleep(10);
        }
        catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        finally {
            this.currentTests.decrementAndGet();
        }
    }

    @Test
    public void test1()
    {
        test();
    }

    @Test
    public void test2()
    {
        test();
    }

    @Test
    public void test3()
    {
        test();
    }

    @Test
    public void test4()
    {
        test();
    }
}
```

```java
@Test(singleThreaded = true)
public class TestClass
        extends BaseTestClass
{
}
```

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="A Suite" parallel="methods" thread-count="4">
    <test name="A test" parallel="methods" thread-count="4">
        <packages>
            <package name="findepi.test.ng" />
        </packages>
    </test>
</suite>
```

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2361 
Closes #144

Fixes #2361  
Fixes #144

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/internal/ClassBasedParallelWorker.java [hunk 1]
```diff
@@ -67,6 +67,10 @@
     return result;
   }
 
+  private static boolean isSame(Set<Class<?>> sequentialClasses, Class<?> c) {
```
Comment (findepi) [inline src/main/java/org/testng/internal/ClassBasedParallelWorker.java:70]: `isChild` ?
or `isSubclass` (with reversed args)?

File: src/test/java/test/thread/issue2361/ChildClassExample.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package test.thread.issue2361;
+
+import org.testng.annotations.Test;
+
+@Test(singleThreaded = true)
+public class ChildClassExample extends BaseTestClassExample {
```
Comment (findepi) [inline src/test/java/test/thread/issue2361/ChildClassExample.java:6]: would a test case using `@Factory` behave any differently?

i experience problem with both the same, but maybe worth asserting both are fixed?

**Comment by findepi** [comment]
Will this solve https://github.com/cbeust/testng/issues/144 too?

(Note that https://github.com/cbeust/testng/issues/144 and https://github.com/cbeust/testng/issues/2361 are similar, yet different)

File: src/main/java/org/testng/internal/ClassBasedParallelWorker.java (hunk 1 above)
Comment (juherr) [reply src/main/java/org/testng/internal/ClassBasedParallelWorker.java:70]: or like an extension `isAssignableFrom`

or `contains` if we consider it like an overriding of the `contains` from Collection

File: src/main/java/org/testng/internal/ClassBasedParallelWorker.java [hunk 3]
```diff
@@ -49,7 +49,7 @@
         params = getParameters(im);
         prevClass = c;
       }
-      if (sequentialClasses.contains(c)) {
+      if (sequentialClasses.contains(c) || isSame(sequentialClasses, c)) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/ClassBasedParallelWorker.java:52]: I think `sequentialClasses.contains(c)` is useless.
If not, prefer to add it into the new method

**Review by juherr — CHANGES_REQUESTED** [review]

File: src/test/java/test/thread/issue2361/ChildClassExample.java (hunk 2 above)
Comment (krmahadevan) [reply src/test/java/test/thread/issue2361/ChildClassExample.java:6]: Fixed

File: src/main/java/org/testng/internal/ClassBasedParallelWorker.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/ClassBasedParallelWorker.java:70]: I renamed the method to `shouldRunSequentially()` and moved the contains logic into it.

File: src/main/java/org/testng/internal/ClassBasedParallelWorker.java (hunk 3 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/ClassBasedParallelWorker.java:52]: `sequentialClasses.contains(c)` is needed. I moved it into the new method.

**Review by juherr — APPROVED** [review]

**Pull request merged**